### ISSUE TITLE
Why "quantized or perfectly correlated input dimensions" may hurt learning?

### ISSUE THREAD

**Issue by ThyrixYang** [VLL-HD/FrEIA#92]
Hi, 

I found in the tips that "Add some slight noise to the inputs (order of 1E-2). This stabilizes training and prevents sparse gradients, if there are some quantized or perfectly correlated input dimensions."

And I noticed that when I repeat the input features (thus "perfectly correlated input"), the log_jac_det will be large, but I have no idea about the reason.

Could you kindly explain this problem?

My practical question is that dimension of input and output is not the same, for example, x.shape=4, y.shape=100. I want to fit a mapping between x and y, but the dimension of the input and output should be the same, how could I achieve this?

**Comment by ThyrixYang** [comment]
Is there any way to reduce dimensions in INNs?  
I know the iRevNet pooling, but the size of the output vectors does not really decrease.

**Comment by MattSkiff** [comment]
It's difficult to give a succinct answer for this - hopefully someone will come along and make this answer redundant.

It intuitively seems obvious it may hinder learning in a similar way to collinearity violating assumptions in statistical models. For one thing, if you have perfectly correlated inputs, it suggests the dimensionality of your input can easily be reduced (i.e. remove features or use dimensionality reduction). Solving an arbitrarily higher dimensionality problem will be more complicated in terms of optimisation (curse of dimensionality). 

I found a casual explanation for how this might hinder gradient descent [here](https://stats.stackexchange.com/questions/191351/what-is-the-connection-between-many-highly-correlated-parameters-in-weight-matri), but I couldn't find any literature on this (though I'm sure it exists).

This [NIPS paper](https://proceedings.neurips.cc/paper/1996/file/884ce4bb65d328ecb03c598409e2b168-Paper.pdf) (which I found from [this Quora answer](https://www.quora.com/In-Neural-Network-if-two-inputs-are-correlated-then-why-would-convergence-of-their-weights-to-the-correct-values-be-slow)) explains in detail how correlated input dimensions can slow down learning using gradient descent.

I am going to interpret sparse gradients as vanishing gradients, although vanishing gradients implies the problem is the architectural choices in the model (such as depth and activation function). One intuitive explanation (don't hold me to this) is that given the initial direction of the gradients on the loss surface are random, there are now more possible incorrect directions to travel in than the lower dimensional 'true' surface you want to learn on (i.e. by finding a low dimension representation of the data and learning on that). Travelling in these incorrect directions causes small gradients, which with a deep architecture or poor choice of activation function could cause vanishing gradients. It also makes sense that if two data sets have the same number of equally informative (independent of each other) dimensions, where input data X has many perfectly correlated dimensions and input data Y doesn't, then input data X will provide a weaker 'signal' than input data Y. That weak signal could manifest as sparse gradients.

This [Quora answer ](https://www.quora.com/What-happens-when-the-inputs-to-a-neural-network-are-highly-correlated) does a simple experiment to show highly correlated data can lead to unstable weights. It should be fairly easy to setup a test of this using synthetic data?
